tests/run/cpp_forwarding_ref.pyx

```diff
@@ -0,0 +1,35 @@
+# mode: run
+# tag: cpp, cpp11
+
+from libcpp.utility cimport move
+
+
+cdef extern from *:
+    """
+    #include <utility>
+
+    const char* f(int& x) {
+        return "lvalue-ref";
+    }
+
+    const char* f(int&& x) {
+        return "rvalue-ref";
+    }
+
+    template <typename T>
+    const char* foo(T&& x)
+    {
+        return f(std::forward<T>(x));
+    }
+    """
+    const char* foo[T](T&& x)
```

Comment:
I know templates are the "hard" bit of this, but it'd probably be good to check that it can handle non-template functions too.

Reply:
Ah apologies -- there may be some confusion here. The intended scope of this PR is handling forwarding references specifically, and not the more general "rvalue references". I should have been clear about the distinction between the two in the description:  

A forwarding reference is defined as a [function parameter of a function template declared as rvalue reference to cv-unqualified type template parameter of that same function template](https://en.cppreference.com/w/cpp/language/reference#Forwarding_references)

Examples:

```c++
template <typename T>
void foo(T&& x);  // x is a forwarding reference

void foo(int&& x);  // x is not a forwarding reference

template <typename T>
void foo(const T&& x);  // x is not a forwarding reference (not const-volatile-unqualified)

template <typename T>
void foo(vector<T>&& x);  // x is not a forwarding reference (not an rvalue reference to the template parameter)
```

So forwarding references by definition appear only in templated functions, and only as input parameters (not return values). They represent a subset of the different contexts in which rvalue references appear in C++.

Forwarding references are different from rvalue references in that they can bind to _both_ lvalues and rvalues. In conjunction with `std::forward`, they are used to "maintain" the lvalue-ness or rvalue-ness of the argument when dispatching to other functions. 

I understand this is a bit of a "backwards" approach, and that adding true support for rvalue references as function parameters and return types in Cython is probably more useful. I implemented this specifically to support https://github.com/cython/cython/pull/3790.

All this being said, I'd be more than happy to look into adding support for `&&` more generally. Although I'd prefer to do that in a separate PR if that's alright.

Reply:
Fair enough - thanks for the clarification. Llimiting it to a small change needed to solve your specific problem makes sense.

It might be worth making sure that users get a helpful error message if they try to do something that isn't currently supported. Put a small file with some of the obvious uses in `tests/errors` (plus a comment to remove the file in future if/when full support is added). Right now this might appear to compile a lot of code, and then not generate sensible results.

Reply:
Will do!

Reply:
Added a few tests.